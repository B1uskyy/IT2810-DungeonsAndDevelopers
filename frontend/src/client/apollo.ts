import { ApolloClient, createHttpLink, InMemoryCache } from '@apollo/client';
import { setContext } from '@apollo/client/link/context';

const httpLink = createHttpLink({
  uri: 'http://localhost:4000/graphql',
});

const authLink = setContext((_, { headers }) => {
  const token = localStorage.getItem('token');
  return {
    headers: {
      ...headers,
      authorization: token ? `Bearer ${token}` : '',
    },
  };
});

const client = new ApolloClient({
  link: authLink.concat(httpLink),
  devtools: {
    enabled: true,
  },
  cache: new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          user: {
            keyArgs: ['id'],
            merge(existing = {}, incoming) {
              return { ...existing, ...incoming };
            },
          },
        },
      },
      User: {
        keyFields: ['id'],
        fields: {
          favoritedMonsters: {
            merge: false,
          },
          dungeonName: {
            merge: false,
          },
          class: {
            merge(_, incoming) {
              return incoming;
            },
          },

          race: {
            merge(_, incoming) {
              return incoming;
            },
          },
        },
      },
      Class: {
        keyFields: ['name'],
      },
      Race: {
        keyFields: ['name'],
      },
      Monster: {
        keyFields: ['id'],
      },
    },
  }),
});

export default client;
