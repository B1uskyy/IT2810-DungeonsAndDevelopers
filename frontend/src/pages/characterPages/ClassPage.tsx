
import Navbar from '../../components/Navbar/Navbar.tsx';
import CharacterButton from '../../components/CharacterCustomization/CharacterButton.tsx';

export default function ClassPage(){
  return (
    <>
      <section className="relative flex items-center justify-center h-screen z-0 before:absolute before:inset-0 before:bg-owlbeast before:bg-cover before:bg-center before:animate-background-zoom  before:z-0">
        <Navbar />
        <section className="w-full h-1/4 relative z-10 flex flex-row items-center justify-around">
          <CharacterButton></CharacterButton>
        </section>
      </section>
      <section className="h-screen w-full bg-gradient-to-b from-owlblue to-black">


      </section>

    </>
  );
}