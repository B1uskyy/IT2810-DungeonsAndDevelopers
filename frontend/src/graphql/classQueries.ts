import { gql } from '@apollo/client';

export const GET_CLASSES = gql`
  query GetClasses($offset: Int, $limit: Int) {
    classes(offset: $offset, limit: $limit) {
      classes {
        id
        index
        name
        hit_die
        skills
      }
      totalClasses
    }
  }
`;
