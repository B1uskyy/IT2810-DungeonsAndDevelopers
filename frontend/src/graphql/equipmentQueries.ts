import { gql } from '@apollo/client';

export const ADD_EQUIPMENT_TO_CHARACTER = gql`
  mutation AddEquipmentToCharacter($userId: ID!, $equipmentId: ID!) {
    addEquipmentToCharacter(userId: $userId, equipmentId: $equipmentId) {
      id
      equipments {
        id
        index
        name
        category
        value
      }
    }
  }
`;

export const REMOVE_EQUIPMENT_FROM_CHARACTER = gql`
  mutation RemoveEquipmentFromCharacter($userId: ID!, $equipmentId: ID!) {
    removeEquipmentFromCharacter(userId: $userId, equipmentId: $equipmentId) {
      id
      equipments {
        id
        index
        name
        category
        value
      }
    }
  }
`;

export const GET_EQUIPMENT_SUGGESTIONS = gql`
  query GetEquipmentSuggestions($searchTerm: String!) {
    equipments(searchTerm: $searchTerm, suggestionsOnly: true) {
      equipments {
        id
        name
      }
      totalCount
    }
  }
`;

export const REMOVE_ALL_EQUIPMENTS = gql`
  mutation RemoveAllEquipments($userId: ID!) {
    removeAllEquipments(userId: $userId) {
      id
      equipments {
        id
        name
        category
        value
      }
    }
  }
`;
export const GET_EQUIPMENTS = gql`
  query GetEquipments($searchTerm: String, $offset: Int, $limit: Int) {
    equipments(searchTerm: $searchTerm, offset: $offset, limit: $limit) {
      equipments {
        id
        index
        name
        category
        value
      }
      totalCount
    }
  }
`;
