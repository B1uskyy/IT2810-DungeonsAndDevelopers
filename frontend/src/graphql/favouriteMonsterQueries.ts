import { gql } from '@apollo/client';

export const ADD_FAVORITE_MONSTER = gql`
  mutation AddFavoriteMonster($userId: ID!, $monsterId: ID!) {
    addFavoriteMonster(userId: $userId, monsterId: $monsterId) {
      favoritedMonsters {
        id
      }
    }
  }
`;

export const REMOVE_FAVORITE_MONSTER = gql`
  mutation RemoveFavoriteMonster($userId: ID!, $monsterId: ID!) {
    removeFavoriteMonster(userId: $userId, monsterId: $monsterId) {
      favoritedMonsters {
        id
      }
    }
  }
`;
