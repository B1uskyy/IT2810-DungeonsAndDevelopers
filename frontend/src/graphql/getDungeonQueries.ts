import { gql } from '@apollo/client';

export const GET_USER_DUNGEON = gql`
  query GetUserDungeon($userId: ID!) {
    user(id: $userId) {
      dungeonName
      favoritedMonsters {
        id
      }
    }
  }
`;
