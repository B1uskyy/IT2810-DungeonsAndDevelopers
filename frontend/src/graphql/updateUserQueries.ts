import { gql } from '@apollo/client';

export const UPDATE_DUNGEON_NAME = gql`
  mutation UpdateDungeonName($userId: ID!, $dungeonName: String!) {
    updateDungeonName(userId: $userId, dungeonName: $dungeonName) {
      id
      dungeonName
    }
  }
`;

export const UPDATE_ABILITY_SCORES = gql`
  mutation UpdateAbilityScores($userId: ID!, $scores: [Int!]!) {
    updateAbilityScores(userId: $userId, scores: $scores) {
      id
      abilityScores {
        ability {
          id
          name
          index
          skills
        }
        score
      }
    }
  }
`;
export const UPDATE_USER_CLASS = gql`
  mutation UpdateUserClass($userId: ID!, $classId: ID!) {
    updateUserClass(userId: $userId, classId: $classId) {
      id
      class {
        id
        name
        index
        hit_die
      }
    }
  }
`;
export const UPDATE_USER_RACE = gql`
  mutation UpdateUserRace($userId: ID!, $raceId: ID!) {
    updateUserRace(userId: $userId, raceId: $raceId) {
      id
      race {
        id
        name
        speed
        alignment
        size
        img
      }
    }
  }
`;
