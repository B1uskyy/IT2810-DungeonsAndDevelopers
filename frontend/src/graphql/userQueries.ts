import { gql } from '@apollo/client';

export const GET_USER_FAVORITES = gql`
  query GetUserFavorites($userId: ID!) {
    user(id: $userId) {
      favoritedMonsters {
        id
        name
        size
        type
        alignment
        hit_points
        image
      }
    }
  }
`;
export const CHECK_USERNAME = gql`
  query checkUsername($userName: String!) {
    checkUsername(userName: $userName)
  }
`;

export const CREATE_USER = gql`
  mutation createUser($userName: String!) {
    createUser(userName: $userName) {
      token
      user {
        id
        userName
        class {
          name
        }
        race {
          name
        }
        abilityScores {
          ability {
            id
            name
          }
          score
        }
        equipments {
          name
        }
      }
    }
  }
`;

export const LOGIN_USER = gql`
  mutation loginUser($userName: String!) {
    loginUser(userName: $userName) {
      token
      user {
        id
        userName
        class {
          name
        }
        race {
          name
        }
        abilityScores {
          ability {
            id
            name
          }
          score
        }
        equipments {
          name
        }
      }
    }
  }
`;
export const GET_USER_CLASS = gql`
  query GetUserClass($userId: ID!) {
    user(id: $userId) {
      class {
        id
        name
        index
        hit_die
      }
    }
  }
`;
export const GET_USER_RACE = gql`
  query GetUserRace($userId: ID!) {
    user(id: $userId) {
      id
      race {
        id
        name
        speed
        alignment
        size
        img
      }
    }
  }
`;
export const GET_ARRAY_SCORES = gql`
  query GetArrayScores($userId: ID!) {
    getArrayScores(userId: $userId) {
      ability {
        id
        name
      }
      score
    }
  }
`;

export const GET_USER_EQUIPMENT = gql`
  query GetUserEquipment($userId: ID!) {
    user(id: $userId) {
      id
      equipments {
        id
        index
        name
        category
        value
      }
    }
  }
`;
