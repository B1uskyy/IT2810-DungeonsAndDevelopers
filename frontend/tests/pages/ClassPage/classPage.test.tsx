import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { render, screen, waitFor, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { GraphQLError } from 'graphql/error';
import { BrowserRouter } from 'react-router-dom';
import { describe, expect } from 'vitest';
import { AuthContext } from '../../../src/context/AuthContext.tsx';
import ClassPage from '../../../src/pages/subPages/classPage.tsx';
import { GET_CLASSES } from '../../../src/graphql/classQueries.ts';
import { GET_USER_CLASS } from '../../../src/graphql/userQueries.ts';
import { UPDATE_USER_CLASS } from '../../../src/graphql/updateUserQueries.ts';

const mockShowToast = vi.fn();

vi.mock('../../../src/hooks/useToast.ts', () => ({
  useToast: () => ({
    showToast: mockShowToast,
  }),
}));

vi.mock('react-responsive', () => ({
  useMediaQuery: () => true, // or false, depending on your needs
}));

const mockUserId = '1';
const mockToken = 'mock-token';
const mockUserName = 'Mock User';

const mockClasses = [
  {
    id: '1',
    index: 'barbarian',
    name: 'Barbarian',
    hit_die: 12,
    skills: [{ name: 'Athletics' }, { name: 'Survival' }],
  },
  {
    id: '2',
    index: 'bard',
    name: 'Bard',
    hit_die: 8,
    skills: [{ name: 'Performance' }, { name: 'Acrobatics' }],
  },
];

const mocks: MockedResponse[] = [
  // Mock for GET_CLASSES
  {
    request: {
      query: GET_CLASSES,
      variables: { offset: 0, limit: 12 },
    },
    result: {
      data: {
        classes: {
          classes: mockClasses,
          totalClasses: mockClasses.length,
        },
      },
    },
  },

  {
    request: {
      query: GET_USER_CLASS,
      variables: { userId: mockUserId },
    },
    result: {
      data: {
        user: {
          id: mockUserId,
          class: mockClasses[0],
        },
      },
    },
  },
];

describe('Classpage component', () => {
  const renderComponent = (additionalMocks: MockedResponse[] = []) => {
    return render(
      <MockedProvider mocks={[...additionalMocks, ...mocks]} addTypename={false}>
        <AuthContext.Provider
          value={{
            userId: mockUserId,
            token: mockToken,
            userName: mockUserName,
            login: vi.fn(),
            logout: vi.fn(),
          }}
        >
          <BrowserRouter>
            <ClassPage />
          </BrowserRouter>
        </AuthContext.Provider>
      </MockedProvider>
    );
  };

  it.skip('renders correctly and matches snapshot', async () => {
    renderComponent();

    await waitFor(() => expect(screen.getByText('Barbarian')).toBeInTheDocument());

    const mainSection = screen.getByRole('main');
    expect(mainSection).toMatchSnapshot();
  });

  it.skip('displays loading state', async () => {
    const loadingMocks = [
      {
        request: {
          query: GET_CLASSES,
          variables: { offset: 0, limit: 12 },
        },
        result: {
          data: null,
        },
        delay: 1000, // test delay
      },
    ];

    renderComponent(loadingMocks);

    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });

  it.skip('displays error state', async () => {
    const consoleErrorMock = vi.spyOn(console, 'error').mockImplementation(() => {});
    const errorMocks = [
      {
        request: {
          query: GET_CLASSES,
          variables: { offset: 0, limit: 12 },
        },
        error: new GraphQLError('Error fetching classes'),
      },
      {
        request: {
          query: GET_USER_CLASS,
          variables: { userId: mockUserId },
        },
        result: {
          data: {
            user: {
              id: mockUserId,
              class: null,
            },
          },
        },
      },
    ];

    renderComponent(errorMocks);
    await waitFor(() => {
      expect(screen.getByText('Error loading classes.')).toBeInTheDocument();
    });

    consoleErrorMock.mockRestore();
  });
  it.skip('updates selected class when a class is clicked', async () => {
    const user = userEvent.setup();

    const userClassMocks = [
      {
        request: {
          query: GET_USER_CLASS,
          variables: { userId: mockUserId },
        },
        result: {
          data: {
            user: {
              id: mockUserId,
              class: null,
            },
          },
        },
      },
      {
        request: {
          query: UPDATE_USER_CLASS,
          variables: { userId: mockUserId, classId: mockClasses[1].id },
        },
        result: {
          data: {
            updateUserClass: {
              id: mockUserId,
              class: mockClasses[1],
            },
          },
        },
      },
    ];

    renderComponent(userClassMocks);

    await waitFor(() => expect(screen.getByText('Barbarian')).toBeInTheDocument());
    const bardCard = (await screen.findByText('Bard')).closest('section') as HTMLElement;
    expect(bardCard).toBeInTheDocument();

    const checkBoxButton = await within(bardCard).findByRole('checkbox');

    await waitFor(() => expect(checkBoxButton).toBeInTheDocument());

    await user.click(checkBoxButton);

    await waitFor(() => {
      expect(checkBoxButton).toBeChecked();
    });
  });
  it.skip('class is already selected', async () => {
    renderComponent();
    await waitFor(() => expect(screen.getByText('Barbarian')).toBeInTheDocument());
    const barbarianCard = (await screen.findByText('Barbarian')).closest('section') as HTMLElement;

    expect(barbarianCard).toBeInTheDocument();

    const checkBox = await within(barbarianCard).findByRole('checkbox');

    await waitFor(() => expect(checkBox).toBeChecked());
  });
});
