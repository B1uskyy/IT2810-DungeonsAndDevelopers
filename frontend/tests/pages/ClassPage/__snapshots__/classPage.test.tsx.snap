// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Classpage component > renders correctly and matches snapshot 1`] = `
<main
  class="relative flex flex-col justify-center items-center w-full bg-subPage bg-cover bg-center min-h-screen"
>
  <div
    class="absolute inset-0 w-full h-full"
  >
    <div
      class="absolute top-0 left-0 w-full h-[10%] bg-black"
    />
    <div
      class="absolute top-[10%] left-0 w-full h-[10%] bg-gradient-to-b from-black to-black/50"
    />
    <div
      class="absolute top-[20%] left-0 w-full h-[80%] bg-black opacity-50"
    />
  </div>
  <section
    class="py-40 w-3/4 flex flex-col items-center gap-36"
  >
    <section
      class="flex w-full h-30 justify-center items-center text-center z-10 space-x-5"
    >
      <div
        class="flex-1 transition-all duration-300 ease-in-out flex items-center justify-center  "
      >
        <a
          class="w-fit relative group flex  header"
          href="/race"
        >
          Race
          <span
            class=""
          />
        </a>
      </div>
      <div
        class="h-14 w-0.5 bg-customRed"
      />
      <div
        class="flex-1 transition-all duration-300 ease-in-out flex items-center justify-center  "
      >
        <a
          class="w-fit relative group flex  header w-full justify-center"
          href="/class"
        >
          Class
          <span
            class=""
          />
        </a>
      </div>
      <div
        class="h-14 w-0.5 bg-customRed"
      />
      <div
        class="flex-1 transition-all duration-300 ease-in-out flex items-center justify-center  "
      >
        <a
          class="w-fit relative group flex  header w-full justify-center"
          href="/abilityscore"
        >
          Ability scores
          <span
            class=""
          />
        </a>
      </div>
    </section>
    <section
      class="flex flex-col items-center w-full gap-10"
    >
      <section
        class="flex flex-col xl:flex-row xl:h-60 w-full justify-between items-center p-8 xl:p-12 rounded-lg bg-black bg-opacity-90 gap-10"
        style="opacity: 0; transform: translateY(100px);"
      >
        <div
          class="flex flex-col xl:flex-row gap-5 justify-center items-center"
        >
          <img
            alt="Barbarian"
            class="max-w-44 shadow-none"
            src="/project2/src/assets/images/class/barbarian.png"
          />
          <div>
            <h2
              class="sub-header bold"
            >
              Barbarian
            </h2>
            <p
              class="text"
            >
              HP: 
              12
            </p>
            <p
              class="text"
            >
              Skills: 
              [object Object], [object Object]
            </p>
          </div>
        </div>
        <div
          class="w-1/5 flex items-center justify-center xl:justify-end"
        >
          <label
            class="relative block cursor-pointer transform transition-transform"
            style="transform: scale(2);"
          >
            <input
              aria-label="equipment-checkbox"
              name="checkbox"
              type="checkbox"
            />
            <span
              aria-label="checkbox"
              class="absolute top-0 left-0 h-6 w-6 rounded bg-gray-300 transition-all duration-300 "
            />
          </label>
        </div>
      </section>
      <section
        class="flex flex-col xl:flex-row xl:h-60 w-full justify-between items-center p-8 xl:p-12 rounded-lg bg-black bg-opacity-90 gap-10"
        style="opacity: 0; transform: translateY(100px);"
      >
        <div
          class="flex flex-col xl:flex-row gap-5 justify-center items-center"
        >
          <img
            alt="Bard"
            class="max-w-44 shadow-none"
            src="/project2/src/assets/images/class/bard.png"
          />
          <div>
            <h2
              class="sub-header bold"
            >
              Bard
            </h2>
            <p
              class="text"
            >
              HP: 
              8
            </p>
            <p
              class="text"
            >
              Skills: 
              [object Object], [object Object]
            </p>
          </div>
        </div>
        <div
          class="w-1/5 flex items-center justify-center xl:justify-end"
        >
          <label
            class="relative block cursor-pointer transform transition-transform"
            style="transform: scale(2);"
          >
            <input
              aria-label="equipment-checkbox"
              name="checkbox"
              type="checkbox"
            />
            <span
              aria-label="checkbox"
              class="absolute top-0 left-0 h-6 w-6 rounded bg-gray-300 transition-all duration-300 "
            />
          </label>
        </div>
      </section>
    </section>
  </section>
</main>
`;
